### PR TITLE
Multi-task binary code prediction

### LINKED ISSUE #40: Training Paradigm: Multi-Task Code Prediction with Retrieval
### The problem with single-task training

MedRAP currently trains the full retrieval-augmented pipeline on a single binary label (ICU mortality). This has two related problems:

1. **Label inefficiency**: the encoder learns clinical structure only from ~1000 mortality labels. It has no prior sense of what patterns in an EHR sequence are predictive of anything — it has to discover this entirely from one outcome.

2. **Task narrowness**: a model shaped by mortality prediction may produce query embeddings that retrieve documents relevant to mortality, but poorly represent patients for whom mortality is not the main clinical concern. Richer tasks → richer representations → better retrieval.

### What we're changing

We replace the single mortality prediction task with **N simultaneous clinical code prediction tasks**, while keeping the **full RAP pipeline active throughout** (encoder → query → retriever → fusion → head). No separate pretraining phase — one training run, N tasks, retriever on.

The core hypothesis: as N grows, the encoder and query projector are forced to learn richer, more general clinical representations, which should improve both prediction quality and retrieval relevance. The scale sweep (N = 1, 2, 5, 10, 25, 50, 100, 1000, 10000) directly tests this.

Mortality can simply be **one of the N tasks**, so we get a mortality prediction signal throughout without needing a separate fine-tune step.

---

## Task Formulations

We implement and compare two ways to define each code-prediction task:

### Binary ("will occur")
For each of N codes: predict whether code X appears in the patient's record after `prediction_time` within a fixed horizon D (e.g. 30 days).
- Labels: `(B, N)` binary tensor — 1 if code appears within horizon, 0 otherwise
- Loss: mean BCE across N tasks, masked for any missing labels

### TTE ("time until")
For each of N codes: predict the distribution over time until code X first appears.
- Labels: `(B, N)` TTE times in days + `(B, N)` event flags (0 = censored)
- Loss: piecewise exponential negative log-likelihood (MOTOR-style), summed across N tasks, handles right-censoring naturally
- Strictly richer than binary — it subsumes the "will occur" question and provides a continuous supervision signal

We run binary first (simpler to debug), then TTE, then compare.

---

### What stays the same
The RAP model architecture is **unchanged**: encoder → query projector → retriever → retrieval encoder → fusion → pooling. The retriever sees the same patient query and returns the same document representations regardless of how many tasks we're predicting.

### What changes

**Head**: `Linear(D, 1)` → `Linear(D, N)` for binary, or `PEANNHead(D, N, P)` for TTE.
`PEANNHead` maps the pooled patient state to `(B, N, P)` positive hazard rates via a linear projection + softplus.

**Loss**: Single BCE → masked multi-task BCE (binary) or masked multi-task piecewise exponential NLL (TTE). In both cases the loss is the mean over all valid `(patient, task)` pairs in the batch.

**Task labels**: Single `boolean_value (B,)` column → `(B, N)` tensor of binary labels or TTE times, loaded from a new multi-task label parquet produced offline.

**DataModule**: Wraps existing `meds_torchdata` and augments each batch with the `(B, N)` label tensors, looked up by `(subject_id, prediction_time)`.

---

## Data Pipeline

An offline script (`scripts/prepare_multi_task_labels.py`) takes:
- The MEDS cohort parquets (full EHR events)
- The existing mortality task label files (to get `prediction_time` anchors per patient)

And produces for each split a parquet with N code-prediction labels per patient, keyed by `(subject_id, prediction_time)`. The top-N codes are selected by frequency across the full cohort (configurable: raw count, or stratified to include rare codes). Mortality is always included as one of the N tasks.

This script runs once per (N, formulation) combination and its outputs are stored on the cluster alongside existing label files.

---

## Config / Experiment Structure

New config groups:

```
training/task:   binary_multitask   |  tte_multitask
training/loss:   binary_multitask   |  tte_multitask
training/head:   linear_multitask   |  peann
```

Key hyperparameters exposed per config:

| Parameter | Values |
|---|---|
| `num_tasks` | 1, 2, 5, 10, 25, 50, 100, 1000, 10000 |
| `task_formulation` | `binary` / `tte` |
| `horizon_days` (binary only) | e.g. 30 |
| `num_pieces` (TTE only) | 3 (small), 5 (medium) |
| `time_boundaries` (TTE only) | e.g. [1.0, 30.0] |
| `code_selection` | `top_k` / `frequency_stratified` |

Each experiment is a single `medrap train` run with a different `num_tasks` override — no new CLI commands needed.

---

## Scale Sweep

N = **1, 2, 5, 10, 25, 50, 100, 1000, 10000**

At N=1 this reduces to the current single-task baseline (mortality only), giving a direct apples-to-apples comparison.

For each N we record:
1. **Val AUROC on mortality** (mortality is always included as one of the N tasks)
2. **Multi-task val loss** (overall and per-task where feasible)
3. **Retrieval quality diagnostics** (already logged: query–doc similarity distributions)

Primary plot: mortality AUROC vs. log(N) for binary and TTE formulations.

---

## Fallback

If neither multi-task formulation improves over N=1 at any scale, we revisit the two-stage pretrain (encoder-only TTE on full cohort) + fine-tune (RAP on mortality) approach from issue #37. That approach has stronger theoretical grounding (larger pretraining cohort, MOTOR precedent) but is more operationally complex.

### PR DESCRIPTION
**Pull request by stepanyanhayk**
Closes #40

## What this adds

Replaces single-task mortality training with **N simultaneous binary code-prediction tasks** ("will code X occur within D days?"), keeping the full RAP pipeline active throughout. N sweeps 1 → 10000 by changing one config override.

### New files
| File | Purpose |
|---|---|
| `scripts/prepare_multi_task_labels.py` | Offline label prep: reads MEDS cohort, computes top-N codes, outputs `(subject_id, prediction_time, task_0 … task_{N-1})` parquets |
| `src/medrap/multitask_datamodule.py` | `MultiTaskMEDSDataset` subclasses `MEDSPytorchDataset` via the `data_class` hook to inject `(B, N)` labels into batches; `MultiTaskMEDSDatamodule` wraps the Lightning datamodule |
| `tests/test_multitask.py` | 13 unit tests covering task, loss, and label I/O |
| `conf/training/task/multitask_binary.yaml` | New task config group |
| `conf/training/loss/multitask_binary_bce.yaml` | New loss config group |
| `conf/training/datamodule/meds_multitask.yaml` | New datamodule config group |

### Modified files
| File | Change |
|---|---|
| `src/medrap/task.py` | Add `MultiTaskBinaryClassificationTask` — masked accuracy metric |
| `src/medrap/losses.py` | Add `MultiTaskBCELoss` — masked mean BCE, NaN entries excluded |
| `src/medrap/__init__.py` | Export new classes |

## How to use

**Step 1 — prepare labels (once per dataset)**
```bash
python scripts/prepare_multi_task_labels.py \
  --meds_cohort_dir  /cluster/path/MEDS_cohort \
  --task_labels_dir  /cluster/path/task_labels/in_hospital_mortality \
  --output_dir       /cluster/path/mt_labels_25 \
  --num_tasks        25 \
  --horizon_days     30
```

**Step 2 — train**
```bash
medrap train \
  encoder=rope \
  training/task=multitask_binary \
  training/loss=multitask_binary_bce \
  training/datamodule=meds_multitask \
  training/task.num_tasks=25 \
  training/datamodule.num_tasks=25 \
  training/datamodule.mt_labels_dir=/cluster/path/mt_labels_25 \
  training/datamodule.config.tensorized_cohort_dir=/cluster/path/tensorized \
  training/datamodule.config.task_labels_dir=/cluster/path/task_labels/in_hospital_mortality \
  training/datamodule.config.max_seq_len=512 \
  output_dir=runs/multitask_25
```

Sweep N by changing `num_tasks`. The head `out_dim` is wired to `training.task.output_dim` automatically via the existing `_train.yaml` binding.

## Tests
117/117 pass (104 pre-existing + 13 new).

### CODE REVIEW

**Comment by stepanyanhayk** [comment]
@copilot review

**Review by zaczw — APPROVED** [review]
